notify/drivers/salesforce/salesforce.py

```diff
@@ -0,0 +1,385 @@
+            'username': self.username,
+            'password': self.password,
+        }
+
+        url = '{}/services/oauth2/token'.format(self.auth_url)
+        response = requests.post(url, verify=None, data=data)
+        response.raise_for_status()
+        return response.json()
+
+    def authenticate(self, **kwargs):
+        if self.organizationId:
+            LOG.debug('self.organizationId={}'.format(self.organizationId))
+            LOG.debug('Auth method = SOAP')
+            return self.authenticate_soap(**kwargs)
+        else:
+            LOG.debug('Auth method = REST')
+            return self.authenticate_rest(**kwargs)
+
+
+class Client(object):
+    def __init__(self, oauth2):
+        self.oauth2 = oauth2
+
+        self.access_token = None
+        self.instance_url = None
+
+    sfdc_base_path = '/services/data/v36.0'
+    sfdc_path = sfdc_base_path + '/sobjects'
+
+    def ticket(self, id, sfdc_path=sfdc_path):
+        try:
+            return self.get(sfdc_path + '/proxyTicket__c/{}'.format(id)).json()
+        except requests.HTTPError:
+            return False
+
+    def create_feeditem(self, data, sfdc_path=sfdc_path):
+        return self.post(sfdc_path + '/FeedItem',
+                         data=json.dumps(data),
+                         headers={"content-type": "application/json"})
+
+    def create_case(self, data, sfdc_path=sfdc_path):
+        return self.post(sfdc_path + '/Case',
+                         data=json.dumps(data),
+                         headers={"content-type": "application/json"})
+
+    def get_case(self, id, sfdc_path=sfdc_path):
+        return self.get(sfdc_path + '/Case/{}'.format(id))
+
+    def update_case(self, id, data, sfdc_path=sfdc_path):
+        return self.patch(sfdc_path + '/Case/{}'.format(id),
+                          data=json.dumps(data),
+                          headers={"content-type": "application/json"})
+
+    def environment(self, id, sfdc_path=sfdc_path):
+        return self.get(sfdc_path + '/Environment__c/{}'.format(id)).json()
+
+    def search(self, query, sfdc_base_path=sfdc_base_path):
+        response = self.get(sfdc_base_path + '/query',
+                            params=dict(q=query)).json()
+        while True:
+            for record in response['records']:
+                yield record
+
+            if response['done']:
+                return
+
+            response = self.get(response['nextRecordsUrl']).json()
+
+    def get(self, url, **kwargs):
+        return self._request('get', url, **kwargs)
+
+    def patch(self, url, **kwargs):
+        return self._request('patch', url, **kwargs)
+
+    def post(self, url, **kwargs):
+        return self._request('post', url, **kwargs)
+
+    def delete(self, url, **kwargs):
+        return self._request('delete', url, **kwargs)
+
+    def delete1(self, url, **kwargs):
+        return self._request('post', url, **kwargs)
+
+    def _request(self, method, url, headers=None, **kwargs):
```

Comment:
headers = kwargs.pop("headers", None)

Otherwise there will be a TypeError if kwargs has "headers" in keys